File: app/backend/api/src/main/java/com/example/sms/domain/model/sales/shipping/Shipping.java
package com.example.sms.domain.model.sales.shipping;

import com.example.sms.domain.model.master.department.Department;
import com.example.sms.domain.model.master.department.DepartmentCode;
import com.example.sms.domain.model.master.employee.Employee;
import com.example.sms.domain.model.master.employee.EmployeeCode;
import com.example.sms.domain.model.master.partner.customer.Customer;
import com.example.sms.domain.model.master.partner.customer.CustomerCode;
import com.example.sms.domain.model.master.product.Product;
import com.example.sms.domain.model.master.product.ProductCode;
import com.example.sms.domain.model.sales.order.*;
import com.example.sms.domain.type.money.Money;
import com.example.sms.domain.type.quantity.Quantity;
import lombok.AllArgsConstructor;
import lombok.NoArgsConstructor;
import lombok.Value;

import java.time.LocalDateTime;

import static org.apache.commons.lang3.Validate.isTrue;

/**
 * 出荷
 */
@Value
@AllArgsConstructor
@NoArgsConstructor(force = true)
public class Shipping {
    OrderNumber orderNumber; // 受注番号
    OrderDate orderDate; // 受注日
    DepartmentCode departmentCode; // 部門コード
    LocalDateTime departmentStartDate; // 部門開始日
    CustomerCode customerCode; // 顧客コード
    EmployeeCode employeeCode; // 社員コード
    DesiredDeliveryDate desiredDeliveryDate; // 希望納期
    String customerOrderNumber; // 客先注文番号
    String warehouseCode; // 倉庫コード
    Money totalOrderAmount; // 受注金額合計
    Money totalConsumptionTax; // 消費税合計
    String remarks; // 備考
    Integer orderLineNumber; // 受注行番号
    ProductCode productCode; // 商品コード
    String productName; // 商品名
    Money salesUnitPrice; // 販売単価
    Quantity orderQuantity; // 受注数量
    TaxRateType taxRate; // 消費税率
    Quantity allocationQuantity; // 引当数量
    Quantity shipmentInstructionQuantity; // 出荷指示数量
    Quantity shippedQuantity; // 出荷済数量
    CompletionFlag completionFlag; // 完了フラグ
    Money discountAmount; // 値引金額
    DeliveryDate deliveryDate; // 納期
    Product product; // 商品
    SalesAmount salesAmount; // 販売価格
    ConsumptionTaxAmount consumptionTaxAmount; // 消費税額
    Department department; // 部門
    Customer customer; // 顧客
    Employee employee; // 社員

    /**
     * 全ての属性を指定してShippingを生成するファクトリメソッド。
     */
    public static Shipping of(
            OrderNumber orderNumber,
            OrderDate orderDate,
            DepartmentCode departmentCode,
            LocalDateTime departmentStartDate,
            CustomerCode customerCode,
            EmployeeCode employeeCode,
            DesiredDeliveryDate desiredDeliveryDate,
            String customerOrderNumber,
            String warehouseCode,
            Money totalOrderAmount,
            Money totalConsumptionTax,
            String remarks,
            Integer orderLineNumber,
            ProductCode productCode,
            String productName,
            Money salesUnitPrice,
            Quantity orderQuantity,
            TaxRateType taxRate,
            Quantity allocationQuantity,
            Quantity shipmentInstructionQuantity,
            Quantity shippedQuantity,
            Money discountAmount,
            DeliveryDate deliveryDate,
            Product product,
            SalesAmount salesAmount,
            ConsumptionTaxAmount consumptionTaxAmount,
            Department department,
            Customer customer,
            Employee employee) {

        isTrue(orderQuantity.getAmount() >= shipmentInstructionQuantity.getAmount(), "受注数量より出荷指示数量が多いです。");
        isTrue(orderQuantity.getAmount() >= shippedQuantity.getAmount(), "受注数量より出荷済数量が多いです。");
        isTrue(shipmentInstructionQuantity.getAmount() >= shippedQuantity.getAmount(), "受注指示数量より出荷済数量が多いです。");

        CompletionFlag status = orderQuantity.getAmount() == shippedQuantity.getAmount() ? CompletionFlag.完了 : CompletionFlag.未完了;

        return new Shipping(
                orderNumber,
                orderDate,
                departmentCode,
                departmentStartDate,
                customerCode,
                employeeCode,
                desiredDeliveryDate,
                customerOrderNumber,
                warehouseCode,
                totalOrderAmount,
                totalConsumptionTax,
                remarks,
                orderLineNumber,
                productCode,
                productName,
                salesUnitPrice,
                orderQuantity,
                taxRate,
                allocationQuantity,
                shipmentInstructionQuantity,
                shippedQuantity,
                status,
                discountAmount,
                deliveryDate,
                product,
                salesAmount,
                consumptionTaxAmount,
                department,
                customer,
                employee
        );
    }
}

File: app/frontend/src/models/sales/shipping.ts
import { PageNationType } from "../../views/application/PageNation";

// 出荷情報
export interface ShippingType {
    orderNumber: string;
    orderDate: string;
    departmentCode: string;
    departmentStartDate: string;
    customerCode: string;
    customerBranchNumber: number;
    employeeCode: string;
    desiredDeliveryDate: string;
    customerOrderNumber: string;
    warehouseCode: string;
    totalOrderAmount: number;
    totalConsumptionTax: number;
    remarks: string;
    orderLineNumber: number;
    productCode: string;
    productName: string;
    salesUnitPrice: number;
    orderQuantity: number;
    taxRate: number;
    allocationQuantity: number;
    shipmentInstructionQuantity: number;
    shippedQuantity: number;
    completionFlag: boolean;
    discountAmount: number;
    deliveryDate: string;
    shippingDate?: string;
    salesAmount: number;
    consumptionTaxAmount: number;
    checked?: boolean;
}

// 出荷検索条件
export interface ShippingCriteriaType {
    orderNumber?: string;
    orderDate?: string;
    departmentCode?: string;
    departmentStartDate?: string;
    customerCode?: string;
    customerBranchNumber?: number;
    employeeCode?: string;
    desiredDeliveryDate?: string;
    customerOrderNumber?: string;
    warehouseCode?: string;
    remarks?: string;
    orderLineNumber?: number;
    productCode?: string;
    productName?: string;
    deliveryDate?: string;
    completionFlag?: boolean;
}

// 出荷リソースをAPIリクエスト用に変換
export const mapToShippingResource = (shipping: ShippingType) => {
    return {
        orderNumber: shipping.orderNumber,
        orderDate: shipping.orderDate,
        departmentCode: shipping.departmentCode,
        departmentStartDate: shipping.departmentStartDate,
        customerCode: shipping.customerCode,
        customerBranchNumber: shipping.customerBranchNumber,
        employeeCode: shipping.employeeCode,
        desiredDeliveryDate: shipping.desiredDeliveryDate,
        customerOrderNumber: shipping.customerOrderNumber,
        warehouseCode: shipping.warehouseCode,
        totalOrderAmount: shipping.totalOrderAmount,
        totalConsumptionTax: shipping.totalConsumptionTax,
        remarks: shipping.remarks,
        orderLineNumber: shipping.orderLineNumber,
        productCode: shipping.productCode,
        productName: shipping.productName,
        salesUnitPrice: shipping.salesUnitPrice,
        orderQuantity: shipping.orderQuantity,
        taxRate: shipping.taxRate,
        allocationQuantity: shipping.allocationQuantity,
        shipmentInstructionQuantity: shipping.shipmentInstructionQuantity,
        shippedQuantity: shipping.shippedQuantity,
        completionFlag: shipping.completionFlag,
        discountAmount: shipping.discountAmount,
        deliveryDate: shipping.deliveryDate,
        salesAmount: shipping.salesAmount,
        consumptionTaxAmount: shipping.consumptionTaxAmount
    };
};

// 検索条件をAPIリクエスト用に変換
export const mapToShippingCriteriaResource = (criteria: ShippingCriteriaType) => {
    const isEmpty = (value: unknown) => value === "" || value === null || value === undefined;
    return {
        ...(isEmpty(criteria.orderNumber) ? {} : { orderNumber: criteria.orderNumber }),
        ...(isEmpty(criteria.orderDate) ? {} : { orderDate: criteria.orderDate }),
        ...(isEmpty(criteria.departmentCode) ? {} : { departmentCode: criteria.departmentCode }),
        ...(isEmpty(criteria.departmentStartDate) ? {} : { departmentStartDate: criteria.departmentStartDate }),
        ...(isEmpty(criteria.customerCode) ? {} : { customerCode: criteria.customerCode }),
        ...(isEmpty(criteria.customerBranchNumber) ? {} : { customerBranchNumber: criteria.customerBranchNumber }),
        ...(isEmpty(criteria.employeeCode) ? {} : { employeeCode: criteria.employeeCode }),
        ...(isEmpty(criteria.desiredDeliveryDate) ? {} : { desiredDeliveryDate: criteria.desiredDeliveryDate }),
        ...(isEmpty(criteria.customerOrderNumber) ? {} : { customerOrderNumber: criteria.customerOrderNumber }),
        ...(isEmpty(criteria.warehouseCode) ? {} : { warehouseCode: criteria.warehouseCode }),
        ...(isEmpty(criteria.remarks) ? {} : { remarks: criteria.remarks }),
        ...(isEmpty(criteria.orderLineNumber) ? {} : { orderLineNumber: criteria.orderLineNumber }),
        ...(isEmpty(criteria.productCode) ? {} : { productCode: criteria.productCode }),
        ...(isEmpty(criteria.productName) ? {} : { productName: criteria.productName }),
        ...(isEmpty(criteria.deliveryDate) ? {} : { deliveryDate: criteria.deliveryDate }),
        // completionFlagはブーリアン値なので、特別な処理を行う
        completionFlag: criteria.completionFlag ?? false
    };
};

// 出荷一覧のページネーション情報
export interface ShippingPageInfoType {
    list: ShippingType[];
    pageNum: number;
    pageSize: number;
    size: number;
    startRow: number;
    endRow: number;
    pages: number;
    prePage: number;
    nextPage: number;
    isFirstPage: boolean;
    isLastPage: boolean;
    hasPreviousPage: boolean;
    hasNextPage: boolean;
    navigatePages: number;
    navigatepageNums: number[];
    navigateFirstPage: number;
    navigateLastPage: number;
    total: number;
}

// 初期値
export const initialShipping: ShippingType = {
    orderNumber: "",
    orderDate: "",
    departmentCode: "",
    departmentStartDate: "",
    customerCode: "",
    customerBranchNumber: 0,
    employeeCode: "",
    desiredDeliveryDate: "",
    customerOrderNumber: "",
    warehouseCode: "",
    totalOrderAmount: 0,
    totalConsumptionTax: 0,
    remarks: "",
    orderLineNumber: 0,
    productCode: "",
    productName: "",
    salesUnitPrice: 0,
    orderQuantity: 0,
    taxRate: 0,
    allocationQuantity: 0,
    shipmentInstructionQuantity: 0,
    shippedQuantity: 0,
    completionFlag: false,
    discountAmount: 0,
    deliveryDate: "",
    salesAmount: 0,
    consumptionTaxAmount: 0,
    checked: false
};

export const initialShippingCriteria: ShippingCriteriaType = {
    orderNumber: "",
    orderDate: "",
    departmentCode: "",
    customerCode: "",
    productCode: "",
    productName: "",
    completionFlag: undefined
};

export const initialShippingPageNation: PageNationType = {
    pageNum: 1,
    pageSize: 10,
    size: 0,
    startRow: 0,
    endRow: 0,
    total: 0,
    pages: 0,
    list: [],
    prePage: 0,
    nextPage: 0,
    isFirstPage: true,
    isLastPage: true,
    hasPreviousPage: false,
    hasNextPage: false,
    navigatePages: 8,
    navigatepageNums: [1],
    navigateFirstPage: 1,
    navigateLastPage: 1
};